features/5-basic-sso/api.md

```diff
@@ -0,0 +1,254 @@
+  - This function returns a skygear user, and an access token of the service, via a delegate.
+- `-[SKYContainer loginWithOAuthProvider:(NSString*)providerID, accessToken:(NSString*)accessToken completion:(void(^)(NSError*, SKYUser*))]`
+  - `accessToken` - Client calls this API if it already has an access token, skygear will try to login directly instead of going through the OAuth flow.
+- `-[SKYContainer associateAccountWithProvider:(NSString*)providerID options:(NSDictionary*)options completion:(void(^)(NSError*, SKYUser*))]`
+  - Add a new auth provider to the user by going through the auth flow
+  - `providerID` - A string that identify the login provider
+  - `options`
+    - uxMode - Either `popup`(default), or `redirect`, popup means in-app-browser (SFSafariViewController/WKWebView) and redirect means Safari.app
+    - clientID
+    - scope
+    - version - FB Client SDK only
+    - cookiePolicy - Google Client SDK only
+  - This function returns a skygear user, and an access token of the new service, via delegate.
+- `-[SKYContainer getOAuthTokensWithCompletion:(void(^)(NSError*, NSDictionary*))]`
+  - Return tokens
+
+        [container getOAuthTokensWithCompletion:^(NSDictionary *tokens){
+          //tokens['com.facebook'] is FB's access token
+        }];
+
+### Platform specific APIs
+
+- `-[SKYContainer loginWithFacebook:(NSDictionary*)options]` and `-[SKYContainer loginWithGoogle:(NSDictionary*)options]`
+  - `options`
+    - uxMode - Either `popup`(default), or `redirect`, popup means in-app-browser (SFSafariViewController/WKWebView) and redirect means Safari.app
+    - clientID
+    - scope
+    - version - FB Client SDK only
+    - cookiePolicy - Google Client SDK only
+    - use3rdPartyClient - whether to use 3rd party client, e.g. Facebook iOS SDK. Default `false`
+  - Returns the user logged in via delegate.
+
+# Changes on API at skygear-server
+
+None.
+
+# Changes on Plugin
+
+## Environment variables
+
+- `UNIQUE_EMAIL_FOR_ACCOUNTS`, boolean
+- `FACEBOOK_CLIENT_SECRET`
+- `GOOGLE_API_KEY`
+- `SSO_ENABLED` e.g `FACEBOOK,GOOGLE`.
+
+## Implementation
+
+The plugin internally has a "Map<ProviderID, (info -> URL)>" map, something like this.
+It will be used to generate urls when user auth.
+
+```
+{
+  'com.facebook': {
+    'auth': () =>
+        "https://www.facebook.com/v2.9/dialog/oauth" +
+        "?client_id=" + clientID +
+        "&redirect_uri=" + generateRedirectURL()
+    'access_token': (code)=>
+        "https://www.facebook.com/v2.9/dialog/access_token" +
+        "?client_id=" + clientID +
+        "&client_secret=" + FACEBOOK_CLIENT_SECRET
+  },
+  'com.google': {
+    'auth': ...,
+    'access_token': ...
+  }
+}
+```
+
+Since the map is in plugin, user should be able to extend it easily.
+
+```
+@oauth_auth('com.custom.website')
+def auth():
+  return "http://example.com/auth"
+
+@oauth_access_token('com.custom.website')
+def access_token(code):
+  return "http://example.com/access_token?code="+code
+```
```

Comment:
If we following authProvider convention, we are using class decorator for group of relating function.

https://docs.skygear.io/guides/advanced/auth-provider/python/